### ISSUE TITLE
how to handle authenticated && !req.isAuthenticated()?

### ISSUE THREAD

**Issue by coolaj86** [ciaranj/connect-auth#120]
Let's say I try to auth against my provider and instead of "Allow", I click "Deny".

I would expect that `authenticated` would be `false` and `req.isAuthenticated()` would also be false.

However, in my example I get back `true` and `false` and I'm not redirected back to the provider's dialog when I make a second attempt. Instead I'm immediately taken to my oauth callback, unable to ever allow.

Am I doing something wrong?

**Comment by ciaranj** [comment]
i'll take a look, incidentally inside your authentication provider implementation 'this.trace' should be available to you rather than console.log which should provide clearer diagnostics to you.

**Comment by ciaranj** [comment]
In the version of your code that I'm looking at, verifyAuthSuccess appears to always call 'self.success'   is that intended?

**Comment by coolaj86** [comment]
Yes and no. As I'm building this out I'm adding more features.

Another bug I'm experiencing is that no matter whether I choose 'allow' or 'deny', I'm always getting a failure (which is ironic because I'm always returning success).

But since I have a failure, I figured why not handle the failure case and then I'll look into why the success case isn't working.

**Comment by ciaranj** [comment]
And when you choose allow/deny that information is being passed back in the
callback to your consumer?

**Comment by coolaj86** [comment]
I'm not sure. I need to go back to the original example and see if the allow/deny fields are hard-coded or if I messed something up as I was editing it. I don't recall ever changing it.

Would you care to chat over skype sometime today? I'm coolaj86, I've added you.

**Comment by ciaranj** [comment]
I'm afraid I'm a bit tied up / busy atm. Is the code you've committed
up-to-date as I don't seem to have any of the login boxes etc. that your
video showed :(

**Comment by coolaj86** [comment]
Yes, I've been testing in Chrome. There may be a bug that affects firefox.

I'll make another video in about an hour and show from git clone to up-and-running what I've done.

**Comment by coolaj86** [comment]
Did you rerun `grunt build`?

**Comment by ciaranj** [comment]
ok, thank you that would allow me to rather more quickly diagnose where it
might be going awry.  Fwiw if req.isAuthenticated() is returning true that
means that there is a 'user' object currently stored in that session?

**Comment by ciaranj** [comment]
nope, never ran that at all!

**Comment by coolaj86** [comment]
I updated the README with the grunt build steps

**Comment by ciaranj** [comment]
You may also want to include some of the steps in here:
http://stackoverflow.com/questions/15511677/installing-npm-module-results-in-command-not-foundfor
those of us not familiar with grunt ;)

**Comment by coolaj86** [comment]
if you git pull the README should have that in there

**Comment by ciaranj** [comment]
I also needed to

cd server/lib

 ln -s ../../blogthing-consumer/server.js consumer.js

 ln -s ../../bookface-provider/server.js provider.js

before  'node bin/demo' would work for me.

 ;)  (sorry)

**Comment by coolaj86** [comment]
Try deleting and re-cloning the project. Something is out of sync with what you have and what I have.

**Comment by ciaranj** [comment]
Cloning has improved things (still got no web forms, but the sym links were
not required) ... incidentally some of the problem could be due to:

```
req.authenticate(['foo'], { scope: ["email", "birthday"] },
```

logAuthentication);

That will be having _weird_ affects with the currently packaged version of
connect-auth.  'scope' in this context means something to connect-auth
(something not very well used or implemented I might add :( )

Try req.authenticate(['foo'], logAuthentication); isntead?
- Cj.

**Comment by coolaj86** [comment]
You were exactly right about that scope option.

Now `allow` and `deny` have the proper action.

When I click deny and then go back to click login again (from the consumer) it pops open the dialog, but skips the oauth process. However, clicking on the login (consumer side) a few more times it will eventually get there.

I think it's a cache bug where the browser assumes the previous redirect will be the current redirect. Perhaps there's a way to expire the redirect immediately so that this doesn't happen.

**Comment by ciaranj** [comment]
Possibly, you should be able to confirm that with wireshark (see if the
HTTP traffic is really leaving the browser or not), if that is an issue you
could always just append a 'cache breaker' to your url to force the browser
to always try.
